scripts/network.js

```diff
@@ -5,6 +5,54 @@ import { Mempool, UTXO } from './mempool.js';
 import { getEventEmitter } from './event_bus.js';
 import { STATS, cStatKeys, cAnalyticsLevel } from './settings.js';
 
+/**
+ * A historical transaction type.
+ * @enum {number}
+ */
+export const HistoricalTxType = {
+    UNKNOWN: 0,
+    STAKE: 1,
+    DELEGATION: 2,
+    UNDELEGATION: 3,
+    RECEIVED: 4,
+    SENT: 5,
+};
+
+/**
+ * A historical transaction
+ */
+export class HistoricalTx {
+    /**
+     * @param {HistoricalTxType} type - The type of transaction.
+     * @param {string} id - The transaction ID.
+     * @param {Array<string>} senders - The list of 'input addresses'.
+     * @param {Array<string>} receivers - The list of 'output addresses'.
+     * @param {boolean} shieldedOutputs - If this transaction contains Shield outputs.
+     * @param {number} time - The block time of the transaction.
+     * @param {number} blockHeight - The block height of the transaction.
+     * @param {number} amount - The amount transacted, in coins.
+     */
+    constructor(
```

Comment:
This should probably take a destructured object

Reply:
A bit of a 'nit of a nit' on my behalf, but, I think this won't be *greatly* useful, for reasons:

- It requires double 'typing' the parameters (of the object to deconstruct) *and* then the internal properties of the Class itself, wastes a bit of space in our files.

- Even with typing the object argument, and then it's properties, VS Code then will not be able to properly auto-fill the types of the Object, in the rare case that a historical TX must be manually constructed (perhaps for the Mempool in the future), see image:
![image](https://github.com/PIVX-Labs/MyPIVXWallet/assets/42538664/298b259c-c46c-4bad-b8f5-5543f2ea6fb0)

- Frequent 'constructing/deconstructing' of objects *does* slightly add pressure to the GC, inflates memory usage, and is generally slower when done en-masse, especially when there'll be thousands of these entries for large wallets. (Not huge enough to TRULY hurt anything, but just something nitty to keep in mind).


All that said... it won't nuke MPW, so if you still prefer it, I'll throw it in.

Reply:
> It requires double 'typing' the parameters (of the object to deconstruct) and then the internal properties of the Class itself, wastes a bit of space in our files.

My editor isn't properly typing the parameters unless I declare the fields manually (No jsdoc required), for example:
```js
class A {
a;
/**
* @param{string} a
*/
constructor(a) {
    this.a = a;
}
}
```
Makes `this.a` a string. If i remove the `a;` declaration, it becomes `any`. This is regardless of how the parameters were passed
> Even with typing the object argument, and then it's properties, VS Code then will not be able to properly auto-fill the types of the Object, in the rare case that a historical TX must be manually constructed (perhaps for the Mempool in the future), see image:

Can you send another image demonstrating what it's supposed to do?
> Frequent 'constructing/deconstructing' of objects does slightly add pressure to the GC, inflates memory usage, and is generally slower when done en-masse, especially when there'll be thousands of these entries for large wallets. (Not huge enough to TRULY hurt anything, but just something nitty to keep in mind).

Surely the modern js engines will optimize this? I'm just guessing, but I doubt it would be impactful

Either way, I think it's fine leaving it like this, because it's only being constructed once

Reply:
> Makes this.a a string. If i remove the a; declaration, it becomes any. This is regardless of how the parameters were passed.

Ah, I see what you mean now - Indeed the class remains dependent on manual typing regardless of the param method used.

> Can you send another image demonstrating what it's supposed to do?

VSC is correctly prompting the current param, during construction (and in the case there's hardcoded `'test'|'test2'|'test3'` types it will also suggest these automatically).
![image](https://github.com/PIVX-Labs/MyPIVXWallet/assets/42538664/573e787d-bd34-4303-a158-49cc984be7e3)

> Surely the modern js engines will optimize this? I'm just guessing, but I doubt it would be impactful.

I had [tested this](https://github.com/PIVX-Labs/MyPIVXWallet/commit/bb6d471aff30249f65e3c37f331cf4c14234b2e2) previously with NobleJS back in the Ancient Days:tm: of MPW, I was tracking down excessive GCs and memory clutter for speeding up VanityGen and was reading deep in to V8 optimisation, on tests discovered that the allocations were from filling the GC with dead objects (each de/structure is a new variable or set-of variables, after all, it can't re-use prior allocated objects), the quirks of JS!